### ISSUE TITLE
Midi-QOL compatibility

### ISSUE THREAD

**Issue by KenGitsIt** [MangoFVTT/fvtt-ready-set-roll-5e#547]
As a DM, I find both Midi-QOL and RSR essential for running my games. I will not run a game without both. Unfortunately, the two modules have never been compatible without customization. In previous releases, I have been able to make them compatible by [modifying only Midi-QOL](https://gitlab.com/tposney/midi-qol/-/issues/1098). But with D&D 4.1.1 and the massive rewrites both modules needed to support activities, I now have to modify both. I'm posting this both to help others who want a DIY way to make the two modules work together, and in the hope that the writers of both modules might choose to adapt/adopt my changes into their base products.

**I have RSR: Why do I want Midi-QOL?**
Midi-QOL has so many automation features that different people will answer this question differently. For me, it is primarily for flags and expirations not supported by the base D&D 5e system. For example, Midi-QOL lets me create a Silvery Barbs effect that will automatically give an actor advantage on its next D20 roll, and then immediately expire.

**I have Midi-QOL: Why do I want RSR?**
With the addition of retroactive advantage/disadvantage/critical buttons, Midi-QOL has closed some of the functionality gap between the two. However, Midi's versions of those buttons don't yet work as well; they provide entirely new rolls, not simply an extra die. Also,  Midi has no retroactive buttons for ability/skill checks or saving throws. Separately, I depend on RSR's ability to separate and apply different damage components, so, for example, I can apply the extra 1d8 from Divine Smite only when hitting a fiend or undead.

**How can I customize them to make the latest versions of each compatible?**
The following instructions have been tested with the most recent versions as of this writing: Foundry 12 build 331, D&D 5e 4.1.1, RSR 3.4.2, and Midi-QOL 12.4.12 (beta). Changes need to be made to three files - two belonging to RSR, and one belonging to Midi-QOL.

**Modify ready-set-roll-5e/src/utils/activity.js**
In the function getDamageFromMessage, change the following code:
```
        const usageConfig = {
            isCritical: message.flags[MODULE_SHORT].isCritical ?? false,
            ammunition: actor.items.get(message.flags[MODULE_SHORT].ammunition),
            scaling: message.flags.dnd5e.scaling ?? 0
        }
```
to this:
```
        const usageConfig = {
            isCritical: message.flags[MODULE_SHORT].isCritical ?? false,
            midiOptions: {isCritical: message.flags[MODULE_SHORT].isCritical ?? false},
            ammunition: actor.items.get(message.flags[MODULE_SHORT].ammunition),
            scaling: message.flags.dnd5e.scaling ?? 0
        }
```
Midi-QOL takes over the damage roll and needs the extra flag to recognize RSR's retroactive critical.

**Modify ready-set-roll-5e/src/utils/chat.js**
1. In the function processChatMessage, change the following code:
```
            if (type == ROLL_TYPE.ACTIVITY && message.isAuthor)
            {
                 ActivityUtility.runActivityActions(message);
            }
```
to this:
```
            if (type == ROLL_TYPE.ACTIVITY && message.isAuthor)
            {
		 message.flags[MODULE_SHORT].processed = true;
            }
```
Midi-QOL separately runs the activity actions. Without this change, rolls will be repeated three times. Note that this is likely a breaking change; without additional logic, RSR will now depend on Midi-QOL to work correctly.

2. In the function _processRetroCritButtonEvent, change the following code:
```
        const rolls = message.rolls.filter(r => r instanceof CONFIG.Dice.DamageRoll);
        const crits = await ActivityUtility.getDamageFromMessage(message);
```
to this:
```
	const allOriginalRolls = message.rolls;
        const rolls = message.rolls.filter(r => r instanceof CONFIG.Dice.DamageRoll);
        const crits = await ActivityUtility.getDamageFromMessage(message);
	message.rolls = allOriginalRolls;
```
When Midi-QOL rerolls the damage, it creates all new rolls. This preserves and restores the old ones.

**Modify midi-qol/src/module/Workflow.js**
In the function displayChatCardWithoutDamageDetail, change the following code:
```
		}
		if (data && this.currentAction === this.WorkflowState_Completed) {
```
to this:
```
		}
                return;
		if (data && this.currentAction === this.WorkflowState_Completed) {
```
This prevents Midi-QOL from creating a duplicate chat card when rolling a retroactive critical from RSR. Note that this is likely a breaking change; without additional logic, Midi-QOL will now depend on RSR to work correctly.

**Comment by KenGitsIt** [comment]
@MangoFVTT I'm happy to help with the ongoing maintenance by posting required changes as needed. Obviously, I can't test proactively before you make changes, but I can at least respond to bug reports, especially if I am alerted to them by being tagged.

In implementing the changes I proposed, I suspect the only modification you will need is to put a check for the presence of midi-qol in the one I flagged as breaking, so that it uses your original logic if midi-qol is not present.

As for the one change that relies on a modification to midi-qol, that is of course outside your control. If we can convince Tim to implement that, then great. If not, then the consequence isn't game-breaking - a duplicate chat card is shown when rolling a retroactive crit.

I appreciate your openness to my suggestions!

**Comment by MangoFVTT** [comment]
@KenGitsIt I've implemented the changes you requested and will roll them into the next release. Do note I haven't at all tested these with midi-qol enabled, so if there are any bugs with both modules you'll need to make further requests. It may be easier for you to simply make fork PRs in that case that I can review.

**Comment by KenGitsIt** [comment]
I have a PR in for the midi-qol side of the changes, but Tim wants to test it before approving it, and to do so, he needs this change to be released. Any chance of moving ahead with a release, or are you waiting for other changes?

**Comment by MangoFVTT** [comment]
@KenGitsIt I'm currently abroad until early January over the Christmas holidays, so I don't have access to my dev environment (hence the delay). I'll push a release once I get back home.

**Comment by KenGitsIt** [comment]
Thank you - and have a great trip!

**Comment by MangoFVTT** [comment]
@KenGitsIt the changes you requested should now be released in 3.4.3 👍
